[libsocialcache] Added SocialImageCache and related database.

antseppa opened the pull request:
Implemented mechanism for caching social media images, such as avatars and images which appear in feeds.
It consists of three classes:

SocialImageDatabase:
  - A database which binds together accountId, image source and cached path.

SocialImageCache:
  - A worker class which loads to-be-cached images on the background and is able to return path to cached version of an image.

SocialMediaCacheItem:

  - A QML-accessible item, which provides path to cached image when image source is given to it.

The idea is that instead of providing original source to QML Image, we pass imageFile property from SocialImageCacheItem to it and
everything else happens on the background. If an account is deleted, buteo purges cached images belogin to that account and images older than given threshold.  

Also modified AbstractImageDownloader to handle URL redirection (Facebook avatars are provided that way).

chriadam commented:
Mssing MER# bug number, also the comment/description should be included as commit message if not already.

chriadam commented on src/lib/socialimagesdatabase.cpp:
```diff
@@ -0,0 +1,448 @@
+    static void clearCachedImages(QSqlQuery &query);
+
+    QList<SocialImage::ConstPtr> queryImages(int accountId,
+                                              const QDateTime &olderThan);
```
indentation

chriadam commented on src/lib/socialimagesdatabase.cpp:
```diff
@@ -0,0 +1,448 @@
+
+    struct {
+        QList<int> purgeAccounts;
+        QStringList removeImages;
```
should this be QPair<accountId, imageUrl> instead of a QStringList?

chriadam commented on src/lib/socialimagesdatabase.cpp:
```diff
@@ -0,0 +1,448 @@
+}
+
+QList<SocialImage::ConstPtr> SocialImagesDatabasePrivate::queryImages(int accountId,
+                                                                        const QDateTime &olderThan)
```
indentation

chriadam commented on src/lib/socialimagesdatabase.cpp:
```diff
@@ -0,0 +1,448 @@
+
+    QList<SocialImage::ConstPtr> data;
+
+    QString queryString = QLatin1String("SELECT accountId, "\
```
no need for line continuation backslashes

chriadam commented on src/lib/socialimagesdatabase.cpp:
```diff
@@ -0,0 +1,448 @@
+                               QDateTime::fromTime_t(query.value(3).toUInt()));  // createdTime
+}
+
+void SocialImagesDatabase::removeImage(const QString &imageUrl)
```
I wonder whether this should require accountId also.  Eg, what if one account's post uses an image, but that same image (url) is only used in another account's post from more than 2 weeks ago, or whatever.  We shouldn't purge the image, in this case, because one account is still using it.  I guess this means more data duplication in theory, but shouldn't be much in practice (ie, composite primary key consisting of accountId+imageUrl)

antseppa replied:
Was thinking about that at first but then came to conclusion that it is not needed that all the related accounst own the data as long as one of them does. That's because:

  - I think it's relatively rare real life case where two accounst access same cached image
  - If it does happen and one account is either removed or purges the image as too old, it will simply be reloaded and assigned to the account that uses it the next time.

So in the end the downside is that we would rarely end up doing what earlier feeds implementation used to do constantly, ie. always downloaded the images.

chriadam commented on src/lib/socialimagesdatabase.cpp:
```diff
@@ -0,0 +1,448 @@
+                  "accountId INTEGER,"
+                  "imageUrl TEXT,"
+                  "imageFile TEXT,"
+                  "createdTime INTEGER)");
```
suggest: PRIMARY KEY (accountId, imageUrl).

chriadam commented on src/qml/generic/socialimagecacheitem.h:
```diff
@@ -0,0 +1,73 @@
+
+class SocialImageDownloader;
+class SocialImageCacheItemPrivate;
+class SocialImageCacheItem : public QObject
```
Hrm.  This could be quite heavy if we have a view with a lot of content, due to all of the signal connections and general QObject slowness.  Do we have a lightweight version of this (ie, a populated struct) which can be used in a model, without requiring QObject instantiation / signal connection / etc in our UI views?

antseppa replied:
Good point, actually. I think I can just expose proper methods from Socialmagecache, listen to it's signals directly, encapsulate that to something like SocialCachedImage.qml in feeds UI and eliminate this class completely.

antseppa replied:
Tried that and it works. Each delegate will still connect to signals from SocialImageCache, but there won't be this class in between.

chriadam commented:
A few comments, in general LGTM.  Good catch regarding redirection.

antseppa commented:
Updated: Fixed indentations, added MER# number and removed obsolete SocialImageCacheItem class.

antseppa commented:
Updated: earlier SocialImageDownloader sent a signal when download succeeded or failed. Delegate items listened to those signals and checked if the file in question was theirs. This led to a lot of unnecessary signals being transmitted. Changed so that downloader now keeps track of ongoing requests and only notifies those delegate items who need to know.

chriadam commented:
LGTM.

antseppa commented:
Updated once more: added "expires" field to allow better cache maintenance control.

chriadam commented on src/lib/socialimagesdatabase.cpp:
```diff
@@ -0,0 +1,520 @@
+    QString queryString = QLatin1String("SELECT accountId, "
+                                        "imageUrl, imageFile, createdTime, expires "
+                                        "FROM images "
+                                        "WHERE accountId = :accountId AND expires < :currentTime");
```
if the user changes their datetime or switches timezone, could that trigger a premature deletion?  I guess it's not an issue, since in that case we'd just redownload the image and it would then get a new expiry time in the new timezone etc...

chriadam commented on src/lib/socialimagesdatabase.cpp:
```diff
@@ -0,0 +1,520 @@
+
+    QList<SocialImage::ConstPtr> data;
+
+    int currentTime = QDateTime::currentDateTime().toTime_t();
```
We should probably do currentDateTimeUtc() instead, and use the UTC timestamp also for the expires value when inserting.

chriadam commented:
LGTM, feel free to ignore those two comments as they're not significant concerns.
